packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -57,24 +58,81 @@ export async function tryHotswapDeployment(
 async function findAllHotswappableChanges(
   stackChanges: cfn_diff.TemplateDiff, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
 ): Promise<HotswapOperation[] | undefined> {
-  const promises = new Array<Promise<ChangeHotswapResult>>();
-  stackChanges.resources.forEachDifference(async (logicalId: string, change: cfn_diff.ResourceDifference) => {
-    promises.push(isHotswappableLambdaFunctionChange(logicalId, change, evaluateCfnTemplate));
+  let foundNonHotswappableChange = false;
+  const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
+
+  // gather the results of the detector functions
+  stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const resourceHotswapEvaluation = isResourceChangeHotswappable(change);
+
+    if (resourceHotswapEvaluation === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+      foundNonHotswappableChange = true;
+    } else if (resourceHotswapEvaluation === ChangeHotswapImpact.IRRELEVANT) {
+      // empty 'if' just for flow-aware typing to kick in...
+    } else {
+      const detectorResults = new Array<Promise<ChangeHotswapResult>>();
+      detectorResults.push(isHotswappableLambdaFunctionChange(logicalId, resourceHotswapEvaluation, evaluateCfnTemplate));
+      detectorResults.push(isHotswappableStateMachineChange(logicalId, resourceHotswapEvaluation, evaluateCfnTemplate));
+
+      promises.push(detectorResults);
```

Comment:
How about just:

```suggestion
      promises.push([
        isHotswappableLambdaFunctionChange(logicalId, resourceHotswapEvaluation, evaluateCfnTemplate),
        isHotswappableStateMachineChange(logicalId, resourceHotswapEvaluation, evaluateCfnTemplate),
    ]);
```

? Seems much simpler 🙂.